core/src/test/scala/neotypes/FutureAsyncSpec.scala

```diff
@@ -0,0 +1,92 @@
+package neotypes
+
+import org.scalatest.{FlatSpec, Matchers}
+
+import scala.concurrent.{ExecutionContext, Future}
+import org.scalatest.concurrent.ScalaFutures
+class FutureAsyncSpec extends FlatSpec with Matchers {
+  implicit val ec = ExecutionContext.Implicits.global
+
+  trait FutureAsyncFixture {
+    var counter = 0
+    var resourceIsOpen = true
+    val fa = Future.successful(1)
+    val f: Int => Future[String] = i => Future.successful(i.toString)
+    val finalizer = { (_: Int, _: Option[Throwable]) => {Future.successful {resourceIsOpen = false}}}
+  }
+
+  "async future" should "execute finalizer" in new FutureAsyncFixture{
+    ScalaFutures.whenReady(Async.futureAsync.guarantee(fa)(f)(finalizer)){
+      w =>
+        w shouldBe "1"
+        resourceIsOpen shouldBe false //resource is closed
+    }
+  }
+
+  "async future" should "execute finalizer when fa fails and return the exception" in new FutureAsyncFixture{
+    override val fa = Future.failed(new Exception("Boom!"))
+
+    ScalaFutures.whenReady(Async.futureAsync.guarantee(fa)(f)(finalizer).failed){
+      ex =>
+        ex.getMessage shouldBe "Boom!"
+        resourceIsOpen shouldBe true //resource is closed
+    }
+  }
+
+  "async future" should "execute finalizer when f fails and return the exception" in new FutureAsyncFixture{
```

Comment:
What is the difference between this test and the previous one?

Reply:
This test will enter the flatMap block after parameter "fa" and the function "f" throws an exception when it is invoked.

Reply:
Ah I see now 👍
It is a bit confusing, I think it would be better that instead of overriding values, we pass the inputs to a helper method.